sys/ztimer/core.c

```diff
@@ -93,9 +140,24 @@ uint32_t ztimer_set(ztimer_clock_t *clock, ztimer_t *timer, uint32_t val)
 
     uint32_t now = _ztimer_update_head_offset(clock);
 
+    bool was_set = false;
     if (_is_set(clock, timer)) {
-        _del_entry_from_list(clock, &timer->base);
+        was_set = _del_entry_from_list(clock, &timer->base);
+    }
+
+#if MODULE_ZTIMER_ONDEMAND
+    /* if the timer wasn't reset on this clock, it acquires the required
+     * ztimer_clock_t. */
+    if (!was_set && _ztimer_acquire(clock) == true) {
```

Comment:
Thank you for looking into the code!

Yes right. But this solution turned out to be less code and complexity - at least when I try to implement it :D

Otherwise I have to do two conditional checks:
1. Is this is first `ztimer_t` in the list?
2. If yes: Is this the first user of the clock.

This results into three different paths through the code:
1. This is the first timer in the list and the first user of the clock
2. This is the first timer in the list but someone already has acquired the clock
3. This isn't the first timer in the list

I'd say the proposed solution only has to different paths:
1. The timer is the first user of the clock
2. The timer isn't the first user of the clock

Or am I missing something?

Reply:
why would the list care if there is any other user?
i see one path:
on set:
timer is first in the list -> list will become a clock user -> increase usercount (aquire handles that this might start the clock)

on remove
if last timer in list -> list will no longer use clock ->  decrease usercount (release will stop clock if its usercount is 0)

Reply:
I'm talking about the two nested if clauses. The three paths are:

1. first if `false`
2. first if `true`, second if `false`
3. first if `true`, second if `true`

I've just pushed ba77e7b3fc8ee7974d00eb783ae82976b511e59f. I tried to move as much work as possible after the `clock->ops->set()` call. It should reduce the cycles through the hot path.